homeassistant/helpers/entity.py

```diff
@@ -772,9 +772,10 @@ def _friendly_name_internal(self) -> str | None:
         ):
             return name
 
+        device_name = device_entry.name_by_user or device_entry.name
```

Comment:
This makes me wonder if we should even allow a device to have no name. Should we make up a random name in the device registry if name is empty?

Reply:
I am not sure about that, it is that with mqtt we have some tests and examples where no device name is set, when then a default entity name is assigned we get a `None ` prefix to the friendly name, and `none_` is added to the 2nd part of the entity id. @emontnemery thinks this is a bug. In this case I would just omit the device part as it does not make sense to add it.

Reply:
Shouldn't we reject the configuration instead? if you want to have your entity name be based on a device but there is no device name, it doesn't work ?

Reply:
It does work, you get an unnamed device. Now allowing this would be breaking, as a device name is optional.

Reply:
Right, but shouldn't we phase in that it's required if `has_entity_name` is set to True?

Reply:
That sounds reasonable to me. If if for MQTT items we do not want to bother the user, we could set `has_entity_name` to `False` when no device name is set and make a device_name required when `has_entity_name` is set to `True`.

Reply:
Setting `has_entity_name` to `False` if no device name is set would be a good alternative:
https://github.com/home-assistant/core/pull/95159/files#r1247719960

Reply:
A case where this will not work is when we want to follow the name of the `device_class`. In that case `has_entity_name` must be set to `True`.